Bugfix falsh_map coverity and flash_map test

nvlsianpu opened the pull request:
- fix for coverity issues (#5739, #5738, #6109).
- fix for test of get sectors API.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
test: subsys: storage: flash_map: fix test of get sectors API

Test of f. flash_area_get_sectors was failing because parameter for
passing capacity of the sectors array was uninitialized.
This patch fix this bug.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
subsys: storage: flash_map: Fix Coverity issues

Loop counter was type of signed int while it was compared
to unsigned lvalue in loop condition.


Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

carlescufi commented:
There is something strange here. Could it be that there is a misconfiguration somewhere so that Coverity sees an empty array of `flash_drivers_map` and `flash_dev`? Coverity runs on some of our test cases so I would just like to check if one of the tests is misconfigured so that flash_map is being included in the build but with empty arrays?

nvlsianpu commented:
FLASH device support might be enabled while non of driver implementation is enabled - think that is the issue.

carlescufi commented:
@nvlsianpu is this something we could fix? i.e. change Kconfig so that it requires a flash driver instead of adding the `if()` statement

nvlsianpu commented:
I can introduce something like FLASH_DRIVER_IS_ENABLED keyword, selected by each driver enable configuration.

carlescufi commented:
@nvlsianpu if that is an option I think it would be a better fix

nvlsianpu committed:
subsys: storage: flash_map: enable only if any flash driver is enable

It was possible to have enable flash module while no flash driver
implementation was enabled. This cause coverity issues and unnecessary
initialization call.

This pat introduce FLASH_HAS_DRIVER_ENABLED Kconfig keyword which is
selected once any flash driver is enabled. flash_map switch its
dependency to this keyword.


Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@carlescufi updated

carlescufi commented:
@nashif as discussed yesterday

galak commented on arch/Kconfig:
```diff
@@ -276,6 +276,12 @@ config FLASH_HAS_PAGE_LAYOUT
 	  This option is enabled when the SoC flash driver supports
 	  retrieving the layout of flash memory pages.
 
+config FLASH_HAS_DRIVER_ENABLED
```
This should live in drivers/flash/Kconfig.  Not sure why FLASH_HAS_PAGE_LAYOUT doesn't live there as well.

nvlsianpu replied:
If I had moved this to flash-Kconfig, the Flash driver options were placed paraller to FLASH menuconfig, not under it as I expected.

galak replied:
What do you mean parallel to it?

nvlsianpu replied:
![image](https://user-images.githubusercontent.com/17042885/37100871-67dc4fc8-2224-11e8-9dd3-b80d925f2f12.png)

nvlsianpu replied:
I found out that have to move the keword declaration before `menuconfig FLASH` for fix picture above.

galak replied:
that makes sense, we should do that.

nvlsianpu replied:
done

galak reviewed (changes requested):
Looks good, we should move the one Kconfig symbol into drivers/flash/Kconfig

nvlsianpu committed:
drivers: flash: move FLASH_HAS_PAGE_LAYOUT Kconfig from arch

Keyword FLASH_HAS_PAGE_LAYOUT is related to flash and should
be declared in its Kconfig.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

carlescufi commented:
Thanks @nvlsianpu
